beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,205 @@
+#
+# We initialize random input and mask tensors, and the ``CustomLinear`` module.
+# Before we run the profiler, we warm-up CUDA to ensure accurate
+# performance benchmarking. We wrap the forward pass of our module in the
+# ``profiler.profile`` context manager. The ``with_stack=True`` parameter appends the
+# file and line number of the operation in the trace.
+#
+
+model = CustomLinear(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.double).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+
+######################################################################
+# Pretty print profiler results
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Finally, we print the profiler results. ``profiler.key_averages``
+# aggregates the results either by input shapes or stack trace events.
+# Grouping by input shapes is useful to identify which tensor shapes
+# contribute to runtime.
+#
+# Here, we use ``group_by_stack_n=5`` which aggregates runtimes by the
+# operation and its traceback (truncated to the most recent 5 events), and
+# display the events in the order they are registered. The table can also
+# be sorted by passing a ``sort_by`` argument (refer to the
+# `docs <https://pytorch.org/docs/stable/autograd.html#profiler>`__ for
+# valid sorting keys).
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Improve memory performance
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# Note that the most expensive operations - in terms of memory and time -
+# are at ``forward (11)`` or where we get the mask indices. Let’s try to
+# tackle the memory consumption first. We can see that the ``.to()``
+# operation consumes 953.67 Mb. ``mask`` is initialized with a
+# ``torch.double`` datatype. Can we reduce the memory footprint by casting
+# it to ``torch.float``?
+#
+
+model = CustomLinear(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.float).cuda()
+
+model(input, mask) #warm up
+
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
+
+######################################################################
+# Improve time performance
+# ^^^^^^^^^^^^^^^^^^^^^^^^^
+# That worked, the memory footprint has halved - but while the time
+# consumed has reduced a bit (previously 283ms), it’s still too high at
+# 194ms. Turns out copying a matrix from CUDA to CPU is pretty expensive!
+# The ``aten::copy_`` operator in ``forward (11)`` copies ``mask`` to CPU
+# so that it can use the NumPy ``argwhere`` function. We could eliminate
+# the need to copy into a NumPy array if we use a ``torch`` function
+# ``nonzero()`` here instead.
+#
+
+class CustomLinearLite(torch.nn.Linear):
```

Comment:
let's maybe, to avoid confusion, call it `MyModule` or smth similar, so it's clear that this is a custom user model code
further, we could also explicitly create a linear module inside the ctor and not inherit from Linear

Reply:
agreed